worlds/witness/entity_hunt.py

```diff
@@ -0,0 +1,231 @@
+
+        assert min(allowance_per_area.values()) >= 0, (
+            f"Somehow, an area had a negative weight when picking hunt entities: {allowance_per_area}"
+        )
+
+        remaining_entities, remaining_entity_weights = [], []
+        for area, eligible_entities in self.ELIGIBLE_ENTITIES_PER_AREA.items():
+            for panel in eligible_entities - self.HUNT_ENTITIES:
+                remaining_entities.append(panel)
+                remaining_entity_weights.append(allowance_per_area[area])
+
+        # I don't think this can ever happen, but let's be safe
+        if sum(remaining_entity_weights) == 0:
+            remaining_entity_weights = [1] * len(remaining_entity_weights)
+
+        return self.random.choices(remaining_entities, weights=remaining_entity_weights, k=amount)
+
+    def _pick_all_hunt_entities(self, total_amount: int):
+        """
+        The core function of the EntityHuntPicker in which all Hunt Entities are picked,
+        respecting the player's choices for total amount and same area discouragement.
+        """
+        same_area_discouragement = self.player_options.panel_hunt_discourage_same_area_factor / 100
+
+        # If we're using random picking, just choose all the entities now and return
+        if not same_area_discouragement:
+            hunt_entities = self.random.sample(
+                [entity for entity in self.ALL_ELIGIBLE_ENTITIES if entity not in self.HUNT_ENTITIES],
+                k=total_amount - len(self.HUNT_ENTITIES),
+            )
+            self.HUNT_ENTITIES.update(hunt_entities)
+            return
+
+        # If we're discouraging entities from the same area being picked, we have to pick entities one at a time
+        # For higher total counts, we do them in small batches for performance
+        batch_size = max(1, total_amount // 20)
+
+        while len(self.HUNT_ENTITIES) < total_amount:
+            actual_amount_to_pick = min(batch_size, total_amount - len(self.HUNT_ENTITIES))
+
+            self.HUNT_ENTITIES.update(self._get_next_random_batch(actual_amount_to_pick, same_area_discouragement))
+
+    def _replace_unfair_hunt_entities_with_good_hunt_entities(self):
+        """
+        For connected entities that "solve together", make sure that the one you're guaranteed
+        to be able to see and interact with first is the one that is chosen, so you don't get "surprise entities".
+        """
+
+        replacements = {
+            "0x18488": "0x00609",  # Replace Swamp Sliding Bridge Underwater -> Swamp Sliding Bridge Above Water
+            "0x03676": "0x03678",  # Replace Quarry Upper Ramp Control -> Lower Ramp Control
+            "0x03675": "0x03679",  # Replace Quarry Upper Lift Control -> Lower Lift Control
+
+            "0x03702": "0x15ADD",  # Jungle Vault Box -> Jungle Vault Panel
+            "0x03542": "0x002A6",  # Mountainside Vault Box -> Mountainside Vault Panel
+            "0x03481": "0x033D4",  # Jungle Vault Box -> Jungle Vault Panel
```

Comment:
Also, shouldn't this include Shipwreck?
```suggestion
            "0x03481": "0x033D4",  # Tutorial Vault Box -> Tutorial Vault Panel
```